OpenRA.Mods.Common/Activities/Air/FlyOffMap.cs

```diff
@@ -11,32 +11,64 @@
 
 using OpenRA.Activities;
 using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Activities
 {
 	public class FlyOffMap : Activity
 	{
 		readonly Aircraft aircraft;
+		readonly Target target;
+		readonly bool hasTarget;
+		bool leftMap;
+		int offMapTicks;
 
 		public FlyOffMap(Actor self)
 		{
 			aircraft = self.Trait<Aircraft>();
+			ChildHasPriority = false;
+		}
+
+		public FlyOffMap(Actor self, Target target)
+		{
+			aircraft = self.Trait<Aircraft>();
+			ChildHasPriority = false;
+			this.target = target;
+			hasTarget = true;
+		}
+
+		protected override void OnFirstRun(Actor self)
+		{
+			if (hasTarget)
+			{
+				QueueChild(new Fly(self, target));
+				return;
+			}
+
+			// VTOLs must take off first if they're not at cruise altitude
+			if (aircraft.Info.VTOL && self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition) != aircraft.Info.CruiseAltitude)
+				QueueChild(new TakeOff(self));
+
+			QueueChild(new FlyTimed(-1, self));
 		}
 
 		public override bool Tick(Actor self)
 		{
 			// Refuse to take off if it would land immediately again.
 			if (aircraft.ForceLanding)
-			{
 				Cancel(self);
+
+			if (IsCanceling)
 				return true;
-			}
 
-			if (IsCanceling || !self.World.Map.Contains(self.Location))
+			if (!leftMap && !self.World.Map.Contains(self.Location))
+				leftMap = true;
+
+			// Wait a few ticks so the aircraft is completely off-map before ending activity
+			if (leftMap && offMapTicks++ > 25)
```

Comment:
IMO it would be good practice to expose the 25 as an argument (which can still default to 25) in the constructor.

Reply:
OK.
But for the record, I tested higher values and for whatever crazy reasons they made the TD C17 circle back into the map before getting removed. And yes, I even disabled the 'nudge aircraft back into map' logic.

I'm too fed up with the aircraft code to debug that further for the forseeable future, though.